src/catalog/src/information_schema/key_column_usage.rs

```diff
@@ -0,0 +1,323 @@
+        catalog_name: String,
+        catalog_manager: Weak<dyn CatalogManager>,
+    ) -> Self {
+        Self {
+            schema,
+            catalog_name,
+            catalog_manager,
+            constraint_catalog: StringVectorBuilder::with_capacity(42),
+            constraint_schema: StringVectorBuilder::with_capacity(42),
+            constraint_name: StringVectorBuilder::with_capacity(42),
+            table_catalog: StringVectorBuilder::with_capacity(42),
+            table_schema: StringVectorBuilder::with_capacity(42),
+            table_name: StringVectorBuilder::with_capacity(42),
+            column_name: StringVectorBuilder::with_capacity(42),
+            ordinal_position: UInt32VectorBuilder::with_capacity(42),
+            position_in_unique_constraint: UInt32VectorBuilder::with_capacity(42),
+            referenced_table_schema: StringVectorBuilder::with_capacity(42),
+            referenced_table_name: StringVectorBuilder::with_capacity(42),
+            referenced_column_name: StringVectorBuilder::with_capacity(42),
+        }
+    }
+
+    /// Construct the `information_schema.KEY_COLUMN_USAGE` virtual table
+    async fn make_key_column_usage(&mut self) -> Result<RecordBatch> {
+        let catalog_name = self.catalog_name.clone();
+        let catalog_manager = self
+            .catalog_manager
+            .upgrade()
+            .context(UpgradeWeakCatalogManagerRefSnafu)?;
+
+        for schema_name in catalog_manager.schema_names(&catalog_name).await? {
+            if !catalog_manager
+                .schema_exists(&catalog_name, &schema_name)
+                .await?
+            {
+                continue;
+            }
+
+            for table_name in catalog_manager
+                .table_names(&catalog_name, &schema_name)
+                .await?
+            {
+                if let Some(table) = catalog_manager
+                    .table(&catalog_name, &schema_name, &table_name)
+                    .await?
+                {
+                    let keys = &table.table_info().meta.primary_key_indices;
+                    let schema = table.schema();
+
+                    for (idx, column) in schema.column_schemas().iter().enumerate() {
+                        let constraint_name = if column.is_time_index() {
+                            "TIME INDEX"
+                        } else if keys.contains(&idx) {
+                            "PRIMARY"
+                        } else {
+                            // TODO: foreign constraint
+                            continue;
+                        };
+
+                        self.add_key_column_usage(
+                            "def",
+                            &schema_name,
+                            constraint_name,
+                            "def",
+                            &schema_name,
+                            &table_name,
+                            &column.name,
+                            idx as u32 + 1,
```

Comment:
The `ordinal_position` is the column's position **within the constraint, not the column's position within the table**.

 So I think we can add a mutable value to index the primary columns just like:

```rust
let keys = &table.table_info().meta.primary_key_indices;
                    let schema = table.schema();

let primary_key_index = 0;

for (idx, column) in schema.column_schemas().iter().enumerate() {
    let constraint_name = if column.is_time_index() {
          "TIME INDEX"
    } else if keys.contains(&idx) {
          primary_key_index = primary_key_index + 1;
          "PRIMARY"
    } else {
           // TODO: foreign constraint
           continue;
    };

    self.add_key_column_usage(
                            "def",
                            &schema_name,
                            constraint_name,
                            "def",
                            &schema_name,
                            &table_name,
                            &column.name,
                            primary_key_index,
                            None,
                            None,
                            None,
                            None,
                        );
}



```

Reply:
Got it.

Should I count one column multiple times when there are multiple key constraints? I find that there is only `time index` and `primary` now, and `time index` is treated as `primary`, do I need to leave a `TODO` here?